src/transformers/models/vision_encoder_decoder/modeling_flax_vision_encoder_decoder.py

```diff
@@ -0,0 +1,835 @@
+            tensors for more detail.
+        output_hidden_states (:obj:`bool`, `optional`):
+            Whether or not to return the hidden states of all layers. See ``hidden_states`` under returned tensors for
+            more detail.
+        return_dict (:obj:`bool`, `optional`):
+            If set to ``True``, the model will return a
+            :class:`~transformers.file_utils.FlaxCausalLMOutputWithCrossAttentions` instead of a plain tuple.
+"""
+
+
+class FlaxVisionEncoderDecoderModule(nn.Module):
+    config: VisionEncoderDecoderConfig
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        encoder_config = self.config.encoder
+        decoder_config = self.config.decoder
+
+        # Copied from `modeling_hybrid_clip.py` with modifications.
+        from ...models.auto.modeling_flax_auto import FLAX_MODEL_FOR_CAUSAL_LM_MAPPING, FLAX_MODEL_MAPPING
+
+        encoder_module = FLAX_MODEL_MAPPING[encoder_config.__class__].module_class
+        decoder_module = FLAX_MODEL_FOR_CAUSAL_LM_MAPPING[decoder_config.__class__].module_class
+
+        self.encoder = encoder_module(encoder_config, dtype=self.dtype)
+        self.decoder = decoder_module(decoder_config, dtype=self.dtype)
+
+        # encoder outputs might need to be projected to different dimension for decoder
+        if (
+            self.encoder.config.hidden_size != self.decoder.config.hidden_size
+            and self.decoder.config.cross_attention_hidden_size is None
+        ):
+            self.enc_to_dec_proj = nn.Dense(
+                self.decoder.config.hidden_size,
+                kernel_init=jax.nn.initializers.normal(self.decoder.config.initializer_range, self.dtype),
+                dtype=self.dtype,
+            )
+        else:
+            self.enc_to_dec_proj = None
+
+    def _get_encoder_module(self):
+        return self.encoder
+
+    def _get_projection_module(self):
+        return self.enc_to_dec_proj
+
+    def _get_decoder_module(self):
+        return self.decoder
+
+    def __call__(
+        self,
+        pixel_values,
+        decoder_input_ids,
+        decoder_attention_mask,
+        decoder_position_ids,
+        output_attentions: bool = False,
+        output_hidden_states: bool = False,
+        return_dict: bool = True,
+        deterministic: bool = True,
+    ):
+        encoder_outputs = self.encoder(
+            pixel_values=pixel_values,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            deterministic=deterministic,
+        )
+
+        encoder_hidden_states = encoder_outputs[0]
+
+        # optionally project encoder_hidden_states
+        if self.enc_to_dec_proj is not None:
+            encoder_hidden_states = self.enc_to_dec_proj(encoder_hidden_states)
+
+        batch_size, sequence_length = encoder_hidden_states.shape[:2]
+        encoder_attention_mask = jnp.ones((batch_size, sequence_length))
```

Comment:
Is this required?

Reply:
Well, currently I take `FlaxGPT2Module` as example, and it could handle the case where the attention mask is `None`.
Does Hugging Face prefers leaving this cross attention mask to the decoder itself, and avoiding pass it explicitly in (Vision)-Encoder-Decoder model?

Reply:
Let me change to `None` and see if everything still works.

Reply:
It works. The test is not green due to some unrelated (tf) issues.

Reply:
Actually, let's please not change it to `None`. If we change it to None it is created later down the road anyway so we don't gain anything from not creating it. The advantage of leaving it as `jnp.ones((batch_size, sequence_length))` is that the TPU XLA compiler knows as soon as possible what shape this variable has and can better optimize. Also passing None can lead to some problems when jitting the model.

In Flax/JAX, we only want to pass `None` for non-tensor function inputs. For all tensor function inputs we should always pass a tensor and not None.

Reply:
OK!

Reply:
@NielsRogge @ydshieh 

As a good rule for Flax/JAX is:

All Flax...Module that inherit from `flax.linen.Module` should **always** require to be passed each tensor argument of their call method. This means, *i.e.* that it's ok to see:

```python
def __call__(
    self, 
    hidden_states: jnp.ndarray,
    output_attentions: Optional[bool] = None,
    ...
)
```

but we don't want to see:

```python
def __call__(
    self, 
    hidden_states: Optional[jnp.ndarray] = None,
    output_attentions: Optional[bool] = None,
    ...
)
```

in the `flax.linen.Module`.

In the `Flax...Model` class this is ok since it's not a `linen.Module` that will be jitted. However inside the `Flax...Model` we should always pass tensors to its attached `linen.Module` class (each `Flax...Model` has a `self.module` property).

In the case above, the `linen.Module` class of the Decoder of FlaxVisionEncoderDecoder should except to get a `encoder_attention_mask` and not None, so let's not pass `None` in this case. We can leave a nice comment like:

```
# make sure encoder_attention_mask for decoder is always initialized to 1 since encoder does not use an attention_mask
```

Also cc @patil-suraj to hear his opinion.

Reply:
@patrickvonplaten Thanks for the information. I have this idea intuitively when I copied the code from other Flax models.

BTW (just for the record), when the tensor arguments are really optional (because models can be used with/without cross-attention, or another arguments like `token_type_ids` or `position_ids`), they are currently being `encoder_attention_mask: Optional[jnp.ndarray] = None,`, for example

https://github.com/huggingface/transformers/blob/a14d62b0b11e6cb0c64059606e4f8dbf78e40e41/src/transformers/models/gpt2/modeling_flax_gpt2.py#L591-L598
or
https://github.com/huggingface/transformers/blob/a14d62b0b11e6cb0c64059606e4f8dbf78e40e41/src/transformers/models/bert/modeling_flax_bert.py#L624-L630

So in this case, the decoder (say `FlaxGPT2Module`) actually expect `encoder_attention_mask: Optional[jnp.ndarray] = None,`.

I can still set it to `1s` in `FlaxVisionEncoderDecoderModule` thought

Reply:
Agree with Patrick!
To add more, I think `None` checks should be fine one or two levels deep for cases such as cross attention where we need `encoder_hidden_states`. For such cases, if input is None then it should stay None throughout. If we need to instantiate it at some point then it's better to create and pass it at the top level. This avoids unnecessary control flow which could (possibly) cause issues with jax transformations.